### PR TITLE
refactor: EntryOptions and EntryPlugin

### PR DESCRIPTION
**Pull request by ahabhgk**
Pre-PR for https://github.com/web-infra-dev/rspack/issues/3552 and https://github.com/web-infra-dev/rspack/issues/2653

### <samp>🤖 Generated by Copilot at f26108a</samp>

This pull request introduces new structs and enums for configuring chunk groups and their entrypoints, and refactors the code that creates and manages them. It also simplifies the chunk id generation logic and removes some unused or redundant code. The main files affected are `code_splitter.rs`, `chunk_group.rs`, `chunk.rs`, `compilation.rs`, and the dependency-related files in `rspack_plugin_javascript`.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at f26108a</samp>

*  Add new structs and methods for configuring chunk groups and their entry options ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e9bb72b14b36e8f0e6c44ef062db46f08893bbdefe39555943ca0d58d00544b0L190-R220), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e9bb72b14b36e8f0e6c44ef062db46f08893bbdefe39555943ca0d58d00544b0L94-R94), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e9bb72b14b36e8f0e6c44ef062db46f08893bbdefe39555943ca0d58d00544b0L103-R103), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e9bb72b14b36e8f0e6c44ef062db46f08893bbdefe39555943ca0d58d00544b0L37-R41), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-a08296f5303f24e761f894dfd29ed0b1afbd4d0c04f56ea364128b94d23183acL11-R11))
*  Update imports to include new structs and methods ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-d3f94114cda1156f8ad0133a36b6e81ae7af80ec628597a73516e92ced5012a8L9-R12), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-86272f0bc7b7b8e6b0ab182b026116a2d16fac8ee7e4de45c4a473aaa33517abL8-R9), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L36-R43), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-65726b45e3f83132a525b4fa2c0f7dc69dc45d35e167175a1e8912bda28d82cfL11-R13), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-bc203fcb806bbd6f59d0f8e10f744904335fdf289d5470b46e60ee5b886370c8L41-R43), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-4d7897467e893502d6e4c09351eed5ff90f9834658555704a8eff4408d6aa8f7L6-R8), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-48492df7cfe0ccf7bce1dbb71175d0019ff15fa50d3510dcf1ff931c38ccaa51L2-R3), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-a40599a3fa2b0310d2e396c186116c51af958a3b60ef9d8fd219d8dc82bab3a5L2-R4), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-8281ab635aae318f63d8e3d743577a933a15bda2136698d2900e8bd8c711d2faL4-R7))
*  Replace `name` parameter and field with `group_options` parameter and field for dependencies that create chunk groups ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-65726b45e3f83132a525b4fa2c0f7dc69dc45d35e167175a1e8912bda28d82cfL24-R26), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-65726b45e3f83132a525b4fa2c0f7dc69dc45d35e167175a1e8912bda28d82cfL54-R59), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-65726b45e3f83132a525b4fa2c0f7dc69dc45d35e167175a1e8912bda28d82cfL98-R100), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-bc203fcb806bbd6f59d0f8e10f744904335fdf289d5470b46e60ee5b886370c8L278-R281), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-bc203fcb806bbd6f59d0f8e10f744904335fdf289d5470b46e60ee5b886370c8L308-R312), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-48492df7cfe0ccf7bce1dbb71175d0019ff15fa50d3510dcf1ff931c38ccaa51L16-R18), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-48492df7cfe0ccf7bce1dbb71175d0019ff15fa50d3510dcf1ff931c38ccaa51L26-R27), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-48492df7cfe0ccf7bce1dbb71175d0019ff15fa50d3510dcf1ff931c38ccaa51L34-R35), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-48492df7cfe0ccf7bce1dbb71175d0019ff15fa50d3510dcf1ff931c38ccaa51L73-R75))
*  Replace `chunk_name` method with `group_options` method for dependencies that create chunk groups ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-4d7897467e893502d6e4c09351eed5ff90f9834658555704a8eff4408d6aa8f7L130-R131), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-4d7897467e893502d6e4c09351eed5ff90f9834658555704a8eff4408d6aa8f7L143-R144))
*  Replace `chunk_name` variable with `group_options` variable and use `ChunkGroupOptions` methods to construct and extract options ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-d3f94114cda1156f8ad0133a36b6e81ae7af80ec628597a73516e92ced5012a8L353-R356), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-d3f94114cda1156f8ad0133a36b6e81ae7af80ec628597a73516e92ced5012a8L394-R397), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-a40599a3fa2b0310d2e396c186116c51af958a3b60ef9d8fd219d8dc82bab3a5L79-R81), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-a40599a3fa2b0310d2e396c186116c51af958a3b60ef9d8fd219d8dc82bab3a5L97-R99), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-8281ab635aae318f63d8e3d743577a933a15bda2136698d2900e8bd8c711d2faL157-R158), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-8281ab635aae318f63d8e3d743577a933a15bda2136698d2900e8bd8c711d2faL174-R175))
*  Remove `id` parameter from `Chunk::new` function and set `id` field to `None` or `name` by default ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-86272f0bc7b7b8e6b0ab182b026116a2d16fac8ee7e4de45c4a473aaa33517abL47-R51), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L296-R298), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L304-R306), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-269c100a13b91351b9f928955094fdd7641e93f1f3062c0de339fd041fb4e763L355-R358), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e14b6b89580c5244c3a00b773aeed57c2ac32a5ff76793ffa4d82eb74047fc05L188-R188))
*  Remove `name` field from `Entrypoint` struct and use `ChunkGroupOptions` to store name ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L48))
*  Add new field `async_entrypoints` to `Compilation` struct to store ukeys of async entrypoint chunk groups ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215R73), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215R130))
*  Add new methods `add_chunk_in_group` and `add_async_entrypoint` to `Compilation` struct to create and connect chunks and chunk groups with given options and runtime ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215R312-R380))
*  Remove `add_entrypoint` method from `Compilation` struct as it is replaced by new methods ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L325))
*  Update `entrypoint` variable initialization with new `ChunkGroupKind::new_entrypoint` function and use `ChunkGroupOptions` to set name ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-d3f94114cda1156f8ad0133a36b6e81ae7af80ec628597a73516e92ced5012a8L68-R75))
*  Update `chunk_group` variable initialization with conditional expression and use `ChunkGroupOptions` to set name and entry options ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-d3f94114cda1156f8ad0133a36b6e81ae7af80ec628597a73516e92ced5012a8L422-R447))
*  Update `is_initial`, `get_runtime_chunk`, and `get_entry_point_chunk` methods of `ChunkGroup` struct to use new `ChunkGroupKind::Entrypoint` variant ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e9bb72b14b36e8f0e6c44ef062db46f08893bbdefe39555943ca0d58d00544b0L94-R94), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e9bb72b14b36e8f0e6c44ef062db46f08893bbdefe39555943ca0d58d00544b0L103-R103), [link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-e9bb72b14b36e8f0e6c44ef062db46f08893bbdefe39555943ca0d58d00544b0L116-R116))
*  Update `is_runtime` method of `Chunk` struct to use new `is_entrypoint` method of `ChunkGroupKind` enum ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-86272f0bc7b7b8e6b0ab182b026116a2d16fac8ee7e4de45c4a473aaa33517abL203-R215))
*  Add new method `get_entry_options` to `Chunk` struct to return entry options of chunk group that contains chunk and is entrypoint ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-86272f0bc7b7b8e6b0ab182b026116a2d16fac8ee7e4de45c4a473aaa33517abR64-R75))
*  Update `get_entry_runtime_chunks` method of `Compilation` struct to include runtime chunks of async entrypoints ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L1039-R1116))
*  Remove unused code that stored dynamic import reason in `chunk_reasons` field of `chunk` variable ([link](https://github.com/web-infra-dev/rspack/pull/3545/files?diff=unified&w=0#diff-d3f94114cda1156f8ad0133a36b6e81ae7af80ec628597a73516e92ced5012a8L403-L405))

</details>

### CODE REVIEW

File: packages/rspack/src/config/defaults.ts [hunk 1]
```diff
@@ -153,6 +155,15 @@ const applyExperimentsDefaults = (experiments: ExperimentsNormalized) => {
 		D(experiments.incrementalRebuild, "make", true);
 		D(experiments.incrementalRebuild, "emitAsset", true);
 	}
+
+	if (
+		cache === false &&
+		experiments.incrementalRebuild &&
+		experiments.incrementalRebuild.make
+	) {
```
Comment (underfin) [inline packages/rspack/src/config/defaults.ts:163]: cc @hardfist @jerrykingxyz

File: crates/rspack_plugin_entry/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    compilation: &mut Compilation,
+    param: &mut MakeParam,
+  ) -> PluginMakeHookOutput {
+    if let Some(state) = compilation.options.get_incremental_rebuild_make_state() && !state.is_first() {
+      return Ok(());
+    }
+    let dependency = Box::new(EntryDependency::new(self.entry_request.clone()));
+    let dependency_id = compilation.module_graph.add_dependency(dependency);
+    compilation.add_entry(dependency_id, self.name.clone(), self.options.clone());
+    param.add_force_build_dependency(dependency_id, None);
```
Comment (jerrykingxyz) [inline crates/rspack_plugin_entry/src/lib.rs:39]: maybe we can just change the param to an array, this case only can add dependency

Comment (ahabhgk) [reply crates/rspack_plugin_entry/src/lib.rs:39]: feel free to refactor in another PR

**Review by underfin — APPROVED** [review]

File: crates/rspack_core/src/build_chunk_graph/code_splitter.rs [hunk 3]
```diff
@@ -419,11 +417,32 @@ impl<'me> CodeSplitter<'me> {
...
-        chunk_name.map(|i| i.to_owned()),
-      );
+
+      let mut chunk_group = if let Some(group_options) = group_options && let Some(entry_options) = &group_options.entry_options {
+        chunk.chunk_reasons.push(format!("AsyncEntrypoint({module_identifier})"));
+        self
+          .remove_parent_modules_context
+          .add_root_chunk(chunk.ukey);
+        let mut entrypoint = ChunkGroup::new(
+          ChunkGroupKind::new_entrypoint(false),
```
Comment (ahabhgk) [inline crates/rspack_core/src/build_chunk_graph/code_splitter.rs:427]: AsyncEntrypoint support

**Pull request merged**